src/mlpack/core/data/load_image.hpp

```
@@ -0,0 +1,150 @@
+ * - PIC (Softimage PIC)
+ * - PNM (PPM and PGM binary only).
+ */
+class LoadImage
+{
+ public:
+  /**
+   * LoadImage default constructor.
+   */
+  LoadImage();
+
+  /**
+   * Instantiate the LoadImage object with the image width, height, channels..
+   *
+   * @param width Matrix width for the output matrix.
+   * @param height Matrix height for the output matrix.
+   * @param channels Matrix channels for the output matrix.
+   */
+  LoadImage(int width, int height, int channels);
+
+  /**
+   * Checks if the given image filename is supported.
+   *
+   * @param filename Name of the image file.
+   * @return Boolean value indicating success if it is an image.
+   */
+  bool IsImageFile(std::string& fileName);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param fileName Name of the image file.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(std::string& fileName, arma::Mat<unsigned char>&& outputMatrix);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param fileName Name of the image file.
+   * @param width Width of the image file.
+   * @param height Height of the image file.
+   * @param channels Channels of the image file.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(std::string& fileName,
+          arma::Mat<unsigned char>&& outputMatrix,
+          int *width,
+          int *height,
+          int *channels);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param files A vector containing names of the image file to be loaded.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(std::vector<std::string>& files,
+    arma::Mat<unsigned char>&& outputMatrix);
+
+  /**
+   * Load the image file into the given matrix.
```

It would probably be good to point out that this fails if `HAS_FILESYSTEM` is not available (I guess, that just means C++17 support is not available?).  And it may be better to throw a `std::runtime_error` instead of returning a bool.

---

Looks with gcc you also have to add ``-lstdc++fs`` when you want filesystem. But since we have this nice build bots we should be able to cover a bunch of different setups.

---

Looks like we could something similar as done here: https://github.com/gulrak/filesystem to check for support:

```c++
#if defined(__cplusplus) && __cplusplus >= 201703L && defined(__has_include) && __has_include(<filesystem>)
#include <filesystem>
namespace fs = std::filesystem;
#else
#include <ghc/filesystem.hpp>
namespace fs = ghc::filesystem;
#endif
```